File: ios/BuddyPlay/Features/Games/Racer/RacerViewModel.swift
import Foundation
import Combine
import BuddyCore

@MainActor
final class RacerViewModel: ObservableObject {

    @Published private(set) var state: RacerState
    @Published var localInput: RacerInput
    @Published var rejectMessage: String?

    let host: Peer
    let guest: Peer
    let localPlayerId: UUID

    private var ticker: Timer?

    init(host: Peer, guest: Peer, localPlayerId: UUID, transport: Transport) {
        self.host = host
        self.guest = guest
        self.localPlayerId = localPlayerId
        self.state = RacerPhysics.initialState(host: host, guest: guest)
        self.localInput = RacerInput(player: localPlayerId, throttle: 0, brake: 0, steering: 0)

        // Mini Racer rejects BLE.
        if transport == .ble {
            self.rejectMessage = "Mini Racer needs Wi-Fi or Hotspot — BLE is too slow for real-time play. Pick a turn-based game (Chess or Dice Kingdom) or switch to Wi-Fi."
        }
    }

    /// Start the host's 30 Hz physics loop. Guests would normally receive
    /// state snapshots from the host instead.
    func startTicking() {
        guard rejectMessage == nil else { return }
        stopTicking()
        ticker = Timer.scheduledTimer(withTimeInterval: 1.0/30.0, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.tickOnce() }
        }
    }

    func stopTicking() {
        ticker?.invalidate()
        ticker = nil
    }

    func setThrottle(_ v: Double) {
        localInput = RacerInput(player: localInput.player, throttle: v, brake: localInput.brake, steering: localInput.steering)
        updateLocalCarInput()
    }
    func setBrake(_ v: Double) {
        localInput = RacerInput(player: localInput.player, throttle: localInput.throttle, brake: v, steering: localInput.steering)
        updateLocalCarInput()
    }
    func setSteering(_ v: Double) {
        localInput = RacerInput(player: localInput.player, throttle: localInput.throttle, brake: localInput.brake, steering: v)
        updateLocalCarInput()
    }

    private func updateLocalCarInput() {
        // Stash the latest input on the local car so the next tick uses it.
        if var car = state.cars[localPlayerId] {
            car.lastInput = localInput
            state.cars[localPlayerId] = car
        }
    }

    private func tickOnce() {
        state = RacerPhysics.tick(state, dtMillis: 33)
    }
}

File: apps/BuddyPlay/android/app/src/main/java/com/americangroupllc/buddyplay/games/racer/RacerViewModel.kt
package com.americangroupllc.buddyplay.games.racer

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.americangroupllc.buddyplay.core.domain.RacerInput
import com.americangroupllc.buddyplay.core.domain.RacerPhysics
import com.americangroupllc.buddyplay.core.domain.RacerState
import com.americangroupllc.buddyplay.core.models.Peer
import com.americangroupllc.buddyplay.core.models.Transport
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class RacerViewModel(host: Peer, guest: Peer, val localPlayerId: String, transport: Transport) : ViewModel() {
    val host = host
    val guest = guest

    var state: RacerState by mutableStateOf(RacerPhysics.initialState(host, guest))
        private set
    var rejectMessage: String? by mutableStateOf(
        if (transport == Transport.BLE)
            "Mini Racer needs Wi-Fi or Hotspot — BLE is too slow for real-time play. Pick a turn-based game (Chess or Dice Kingdom) or switch to Wi-Fi."
        else null
    )
        private set

    private var ticker: Job? = null
    private var localInput = RacerInput(localPlayerId, throttle = 0.0, brake = 0.0, steering = 0.0)

    fun startTicking() {
        if (rejectMessage != null) return
        stopTicking()
        ticker = viewModelScope.launch {
            while (true) {
                delay(33)
                state = RacerPhysics.tick(state, dtMillis = 33)
            }
        }
    }

    fun stopTicking() {
        ticker?.cancel(); ticker = null
    }

    fun setThrottle(v: Double) { localInput = localInput.copy(throttle = v); pushInput() }
    fun setBrake(v: Double)    { localInput = localInput.copy(brake = v);    pushInput() }
    fun setSteering(v: Double) { localInput = localInput.copy(steering = v); pushInput() }

    private fun pushInput() {
        val car = state.cars[localPlayerId] ?: return
        state.cars[localPlayerId] = car.copy(lastInput = localInput)
    }
}
